dtm0/service.c

```diff
@@ -0,0 +1,625 @@
+static int dtm0_service_start(struct m0_reqh_service *service);
+static void dtm0_service_stop(struct m0_reqh_service *service);
+static int dtm0_service_allocate(struct m0_reqh_service **service,
+				 const struct m0_reqh_service_type *stype);
+static void dtm0_service_fini(struct m0_reqh_service *service);
+
+
+static const struct m0_reqh_service_type_ops dtm0_service_type_ops = {
+	.rsto_service_allocate = dtm0_service_allocate
+};
+
+static const struct m0_reqh_service_ops dtm0_service_ops = {
+	.rso_start = dtm0_service_start,
+	.rso_stop  = dtm0_service_stop,
+	.rso_fini  = dtm0_service_fini
+};
+
+struct m0_reqh_service_type dtm0_service_type = {
+	.rst_name  = "M0_CST_DTM0",
+	.rst_ops   = &dtm0_service_type_ops,
+	.rst_level = M0_RS_LEVEL_LATE,
+};
+
+
+/**
+ * System process which dtm0 subscribes for state updates
+ */
+struct dtm0_process {
+	struct m0_tlink         dop_link;
+	uint64_t                dop_magic;
+
+	/**
+	 * Listens for an event on process conf object's HA channel.
+	 * Updates dtm0_process status in the clink callback on HA notification.
+	 */
+	struct m0_clink         dop_ha_link;
+	/**
+	 * Link connected to remote process
+	 */
+	struct m0_rpc_link      dop_rlink;
+	/**
+	 * Remote process fid
+	 */
+	struct m0_fid           dop_rproc_fid;
+	/**
+	 * Remote service fid
+	 */
+	struct m0_fid           dop_rserv_fid;
+	/**
+	 * Remote process endpoint
+	 */
+	const char             *dop_rep;
+	/**
+	 * Current dtm0 service dtm0 process to.
+	 */
+	struct m0_reqh_service *dop_dtm0_service;
+	/**
+	 * Connect ast
+	 */
+	struct m0_sm_ast        dop_service_connect_ast;
+	struct m0_clink         dop_service_connect_clink;
+};
+
+M0_TL_DESCR_DEFINE(dopr, "dtm0_process", static, struct dtm0_process, dop_link,
+		   dop_magic, 0x8888888888888888, 0x7777777777777777);
+M0_TL_DEFINE(dopr, static, struct dtm0_process);
+
+/**
+ * typed container_of
+ */
+enum { DTM0_SERVICE_MAGIX = 0x9999999999999999 };
+static const struct m0_bob_type dtm0_service_bob = {
+	.bt_name = "dtm0 service",
+	.bt_magix_offset = M0_MAGIX_OFFSET(struct m0_dtm0_service, dos_magix),
+	.bt_magix = DTM0_SERVICE_MAGIX,
+	.bt_check = NULL
+};
+M0_BOB_DEFINE(static, &dtm0_service_bob, m0_dtm0_service);
+
+static struct m0_dtm0_service *to_dtm(struct m0_reqh_service *service)
+{
+	return bob_of(service, struct m0_dtm0_service, dos_generic,
+		      &dtm0_service_bob);
+}
+
+/**
+ * Service part
+ */
+static int dtm0_service__init(struct m0_dtm0_service *s)
+{
+	m0_dtm0_service_bob_init(s);
+	dopr_tlist_init(&s->dos_processes);
+	m0_dtm0_dtx_domain_init();
+
+	return m0_dtm0_clk_src_init(&s->dos_clk_src, M0_DTM0_CS_PHYS);
+}
+
+static void dtm0_service__fini(struct m0_dtm0_service *s)
+{
+	m0_dtm0_clk_src_fini(&s->dos_clk_src);
+	m0_dtm0_dtx_domain_fini();
+	dopr_tlist_fini(&s->dos_processes);
+	m0_dtm0_service_bob_fini(s);
+}
+
+static struct dtm0_process *dtm0_service_process__lookup(struct m0_reqh_service *reqh_dtm0_svc,
```

Comment:
looks too bloody...


let's try this one:
static struct dtm0_process *
dtm0_service_process__lookup(struct m0_reqh_service *reqh_dtm0_svc,
			     const struct m0_fid *remote_dtm0)
{
	return m0_tl_find(dopr, proc, &to_dtm(reqh_dtm0_svc)->dos_processes,
			  m0_fid_eq(&proc->dop_rserv_fid, remote_dtm0));
}

Reply:
Done.